cpu/kinetis_common/uart.c

```diff
@@ -0,0 +1,323 @@
+    //dev->C4 = (uint8_t)UART_C4_BRFA(brfa);
+
+    /* enable transmitter and receiver */
+    dev->C2 |= (1 << UART_C2_TE_SHIFT | 1 << UART_C2_RE_SHIFT);
+    return 0;
+}
+
+void uart_tx_begin(uart_t uart)
+{
+    switch (uart) {
+#if UART_0_EN
+
+        case UART_0:
+            UART_0_DEV->C2 |= (1 << UART_C2_TIE_SHIFT);
+            break;
+#endif
+#if UART_1_EN
+
+        case UART_1:
+            UART_1_DEV->C2 |= (1 << UART_C2_TIE_SHIFT);
+            break;
+#endif
+
+        default:
+            break;
+    }
+}
+
+void uart_tx_end(uart_t uart)
+{
+    switch (uart) {
+#if UART_0_EN
+
+        case UART_0:
+            UART_0_DEV->C2 &= ~(1 << UART_C2_TIE_SHIFT);
+            break;
+#endif
+#if UART_1_EN
+
+        case UART_1:
+            UART_1_DEV->C2 &= ~(1 << UART_C2_TIE_SHIFT);
+            break;
+#endif
+
+        default:
+            break;
+    }
+}
+
+int uart_write(uart_t uart, char data)
+{
+    switch (uart) {
+#if UART_0_EN
+
+        case UART_0:
+            if (UART_0_DEV->S1 & UART_S1_TDRE_MASK) {
+                UART_0_DEV->D = (uint8_t)data;
+            }
+
+            break;
+#endif
+#if UART_1_EN
+
+        case UART_1:
+            if (UART_1_DEV->S1 & UART_S1_TDRE_MASK) {
+                UART_1_DEV->D = (uint8_t)data;
+            }
+
+            break;
+#endif
+
+        default:
+            return -2;
+            break;
+    }
+
+    return 0;
+}
+
+int uart_read_blocking(uart_t uart, char *data)
+{
+    switch (uart) {
+#if UART_0_EN
+
+        case UART_0:
+            while (!(UART_0_DEV->S1 & UART_S1_RDRF_MASK));
+
+            *data = (char)UART_0_DEV->D;
+            break;
+#endif
+#if UART_1_EN
+
+        case UART_1:
+            while (!(UART_1_DEV->S1 & UART_S1_RDRF_MASK));
+
+            *data = (char)UART_1_DEV->D;
+            break;
+#endif
+
+        default:
+            return -2;
+            break;
+    }
+
+    return 1;
+}
+
+int uart_write_blocking(uart_t uart, char data)
+{
+    switch (uart) {
+#if UART_0_EN
+
+        case UART_0:
+            while (!(UART_0_DEV->S1 & UART_S1_TDRE_MASK));
+
+            UART_0_DEV->D = (uint8_t)data;
+            break;
+#endif
+#if UART_1_EN
+
+        case UART_1:
+            while (!(UART_1_DEV->S1 & UART_S1_TDRE_MASK));
+
+            UART_1_DEV->D = (uint8_t)data;
+            break;
+#endif
+
+        default:
+            return -2;
+            break;
+    }
+
+    return 1;
+}
+
+#if UART_0_EN
+void UART_0_ISR(void)
+{
+    irq_handler(UART_0, UART_0_DEV);
+}
+#endif
+
+#if UART_1_EN
+void UART_1_ISR(void)
+{
+    irq_handler(UART_1, UART_1_DEV);
+}
+#endif
+
+static inline void irq_handler(uint8_t uartnum, UART_Type *dev)
+{
+    if (dev->S1 & UART_S1_RDRF_MASK) {
+        char data = (char)dev->D;
+        config[uartnum].rx_cb(config[uartnum].arg, data);
+    }
+    else if (dev->S1 & UART_S1_TDRE_MASK) {
+        if (config[uartnum].tx_cb(config[uartnum].arg) == 0) {
+            dev->C2 |= (1 << UART_C2_TIE_SHIFT);
```

Comment:
The logic seem wrong. TX interrupt is supposed to be disabled if return value is zero. `dev->C2 &= ~(1 << UART_C2_TIE_SHIFT);` (or shorter: `dev->C2 &= ~(UART_C2_TIE_MASK);`)

See also: http://riot-os.org/api/group__driver__periph__uart.html#ga2aa1686d28caae1aa74baa3d047e8330